src/web/metrics.rs

```diff
@@ -197,6 +204,47 @@ impl iron::Handler for RequestRecorder {
     }
 }
 
+struct RenderingTime {
+    start: Instant,
+    step: &'static str,
+}
+
+pub(crate) struct RenderingTimesRecorder {
+    metric: &'static HistogramVec,
+    current: Option<RenderingTime>,
+}
+
+impl RenderingTimesRecorder {
+    pub(crate) fn new(metric: &'static HistogramVec) -> Self {
+        Self {
+            metric,
+            current: None,
+        }
+    }
+
+    pub(crate) fn step(&mut self, step: &'static str) {
+        self.record_current();
+        self.current = Some(RenderingTime {
+            start: Instant::now(),
+            step,
+        });
+    }
+
+    fn record_current(&mut self) {
+        if let Some(current) = self.current.take() {
```

Comment:
This doesn't count the time to record the metric, is that intentional? If not you could make it (mostly) atomic like this:
```suggestion
        if let Some(current) = self.current.replace(Instant::now()) {
```
and then removing `current = ...` in `step()`.

Reply:
Is that really needed? The thing I want to get out of this metrics is how much it takes to do database queries, fetch files from S3 and parse the HTML, and the time it takes to record metrics won't influence that.

Reply:
Ok, if it's not valuable we don't need to include it.